File: A-Dex/discord-bot/src/commandParser.js
const REMOTE_COMMANDS = new Set([
  'apps',
  'open',
  'lock',
  'say',
  'sayurdu',
  'playaudio',
  'stopaudio',
  'pauseaudio',
  'resumeaudio',
  'audiostatus',
  'parentpin',
  'shield',
  'screenshot',
  'files',
  'filestat',
  'mkdir',
  'rename',
  'move',
  'delete',
  'uploadfile',
  'readtext',
  'download',
  'volume',
  'info',
  'permstatus',
  'location',
  'camerasnap',
  'contactlookup',
  'smsdraft',
  'fileshareintent',
  'quicklaunch',
  'torchpattern',
  'ringtoneprofile',
  'screentimeoutset',
  'mediacontrol',
  'randomquote',
  'fakecallui',
  'shakealert',
  'show',
  'message',
  'lockapp',
  'unlockapp',
  'lockedapps',
  'usage',
]);

const ADMIN_COMMANDS = new Set(['pair', 'bind', 'unbind', 'admins', 'devices']);

// Tokenizer supports quoted arguments for message and say commands.
function tokenize(text) {
  const regex = /"([^"]+)"|'([^']+)'|(\S+)/g;
  const tokens = [];
  let match;
  while ((match = regex.exec(text)) !== null) {
    tokens.push(match[1] || match[2] || match[3]);
  }
  return tokens;
}

function parseCommandInput(content, prefix) {
  if (!content || !content.startsWith(prefix)) {
    return null;
  }

  const raw = content.slice(prefix.length).trim();
  if (!raw) {
    return null;
  }

  const tokens = tokenize(raw);
  if (tokens.length === 0) {
    return null;
  }

  const name = tokens[0].toLowerCase();
  const args = tokens.slice(1);

  return {
    name,
    args,
    isRemote: REMOTE_COMMANDS.has(name),
    isAdmin: ADMIN_COMMANDS.has(name),
  };
}

function buildRemotePayload(name, args, attachment, maxImageBytes) {
  switch (name) {
    case 'apps':
    case 'lock':
    case 'stopaudio':
    case 'pauseaudio':
    case 'resumeaudio':
    case 'audiostatus':
    case 'screenshot':
    case 'files':
    case 'camerasnap':
    case 'randomquote':
    case 'info':
    case 'permstatus':
    case 'location':
    case 'lockedapps':
    case 'usage':
      return { payload: {} };
    case 'open':
      if (!args[0]) {
        return { error: 'Usage: !open <app package or display name>' };
      }
      return { payload: { target: args.join(' ') } };
    case 'say':
      if (args.length === 0) {
        return { error: 'Usage: !say <text>' };
      }
      return { payload: { text: args.join(' ') } };
    case 'sayurdu':
      if (args.length === 0) {
        return { error: 'Usage: !sayurdu <urdu text>' };
      }
      return { payload: { text: args.join(' ') } };
    case 'playaudio': {
      if (!args[0]) {
        return { error: 'Usage: !playaudio <url> [repeat]' };
      }
      let repeat = 1;
      if (args[1] !== undefined) {
        repeat = Number(args[1]);
      }
      if (!Number.isFinite(repeat) || repeat < 1 || repeat > 100) {
        return { error: 'Usage: !playaudio <url> [repeat]' };
      }
      return { payload: { url: args[0], repeat: Math.round(repeat) } };
    }
    case 'parentpin':
      if (!args[0]) {
        return { error: 'Usage: !parentpin <4-12 digit pin>' };
      }
      return { payload: { pin: args[0] } };
    case 'shield': {
      const action = (args[0] || 'status').toLowerCase();
      if (!['enable', 'disable', 'status', 'relock'].includes(action)) {
        return { error: 'Usage: !shield <enable|disable|status|relock>' };
      }
      return { payload: { action } };
    }
    case 'download':
      if (!args[0]) {
        return { error: 'Usage: !download <path>' };
      }
      return { payload: { path: args.join(' ') } };
    case 'filestat':
      if (!args[0]) {
        return { error: 'Usage: !filestat <path>' };
      }
      return { payload: { path: args.join(' ') } };
    case 'mkdir':
      if (!args[0]) {
        return { error: 'Usage: !mkdir <path>' };
      }
      return { payload: { path: args.join(' ') } };
    case 'rename':
      if (!args[0] || !args[1]) {
        return { error: 'Usage: !rename <path> <new_name>' };
      }
      return { payload: { path: args[0], new_name: args[1] } };
    case 'move':
      if (!args[0] || !args[1]) {
        return { error: 'Usage: !move <source> <target_dir>' };
      }
      return { payload: { source: args[0], target_dir: args[1] } };
    case 'delete':
      if (!args[0]) {
        return { error: 'Usage: !delete <path> [recursive:true|false]' };
      }
      return { payload: { path: args[0], recursive: args[1] === 'true' } };
    case 'uploadfile':
      if (!args[0] || !args[1]) {
        return { error: 'Usage: !uploadfile <target_dir> <url> [file_name]' };
      }
      return { payload: { target_dir: args[0], url: args[1], file_name: args[2] } };
    case 'readtext': {
      if (!args[0]) {
        return { error: 'Usage: !readtext <path> [max_chars]' };
      }
      let maxChars = 2000;
      if (args[1] !== undefined) {
        maxChars = Number(args[1]);
      }
      if (!Number.isFinite(maxChars) || maxChars < 1) {
        return { error: 'Usage: !readtext <path> [max_chars]' };
      }
      return { payload: { path: args[0], max_chars: Math.round(maxChars) } };
    }
    case 'volume': {
      const value = Number(args[0]);
      if (!Number.isFinite(value) || value < 0 || value > 100) {
        return { error: 'Usage: !volume <0-100>' };
      }
      return { payload: { value: Math.round(value) } };
    }
    case 'show': {
      const seconds = Number(args[0]);
      if (!Number.isFinite(seconds) || seconds < 1 || seconds > 60) {
        return { error: 'Usage: !show <1-60> with an image attachment' };
      }

      if (!attachment) {
        return { error: 'Attach a single image to use !show' };
      }

      const isImage = (attachment.contentType || '').startsWith('image/');
      if (!isImage) {
        return { error: 'Attachment must be an image' };
      }

      if (attachment.size && attachment.size > maxImageBytes) {
        return { error: `Attachment too large; max ${Math.floor(maxImageBytes / (1024 * 1024))} MB` };
      }

      return {
        payload: {
          seconds: Math.round(seconds),
          imageUrl: attachment.url,
          imageName: attachment.name || 'image',
          imageContentType: attachment.contentType || 'image/*',
        },
      };
    }
    case 'message':
      if (args.length === 0) {
        return { error: 'Usage: !message <text>' };
      }
      return { payload: { text: args.join(' ') } };
    case 'contactlookup': {
      if (!args[0]) {
        return { error: 'Usage: !contactlookup <query> [limit]' };
      }
      const payload = { query: args[0] };
      if (args[1] !== undefined) {
        const limit = Number(args[1]);
        if (!Number.isFinite(limit) || limit < 1) {
          return { error: 'Usage: !contactlookup <query> [limit]' };
        }
        payload.limit = Math.round(limit);
      }
      return { payload };
    }
    case 'smsdraft':
      if (!args[0] || args.length < 2) {
        return { error: 'Usage: !smsdraft <number> <message>' };
      }
      return { payload: { number: args[0], message: args.slice(1).join(' ') } };
    case 'fileshareintent':
      if (!args[0]) {
        return { error: 'Usage: !fileshareintent <path> [mime_type]' };
      }
      return { payload: { path: args[0], mimeType: args[1] } };
    case 'quicklaunch':
      if (!args[0]) {
        return { error: 'Usage: !quicklaunch <package_or_url>' };
      }
      if (args[0].startsWith('http://') || args[0].startsWith('https://')) {
        return { payload: { url: args[0] } };
      }
      return { payload: { packageName: args[0] } };
    case 'torchpattern': {
      const repeats = args[0] === undefined ? 3 : Number(args[0]);
      const onMs = args[1] === undefined ? 250 : Number(args[1]);
      const offMs = args[2] === undefined ? 250 : Number(args[2]);
      if (!Number.isFinite(repeats) || !Number.isFinite(onMs) || !Number.isFinite(offMs)) {
        return { error: 'Usage: !torchpattern [repeats] [on_ms] [off_ms]' };
      }
      return { payload: { repeats: Math.round(repeats), on_ms: Math.round(onMs), off_ms: Math.round(offMs) } };
    }
    case 'ringtoneprofile':
      if (!['normal', 'vibrate', 'silent'].includes((args[0] || '').toLowerCase())) {
        return { error: 'Usage: !ringtoneprofile <normal|vibrate|silent>' };
      }
      return { payload: { mode: args[0].toLowerCase() } };
    case 'screentimeoutset': {
      const seconds = Number(args[0]);
      if (!Number.isFinite(seconds) || seconds < 1) {
        return { error: 'Usage: !screentimeoutset <seconds>' };
      }
      return { payload: { seconds: Math.round(seconds) } };
    }
    case 'mediacontrol':
      if (!['play', 'pause', 'next', 'previous', 'stop', 'toggle'].includes((args[0] || '').toLowerCase())) {
        return { error: 'Usage: !mediacontrol <play|pause|next|previous|stop|toggle>' };
      }
      return { payload: { action: args[0].toLowerCase() } };
    case 'fakecallui':
      return { payload: { callerName: args.join(' ') || 'Unknown Caller' } };
    case 'shakealert': {
      const action = (args[0] || 'status').toLowerCase();
      if (!['start', 'stop', 'status'].includes(action)) {
        return { error: 'Usage: !shakealert <start|stop|status>' };
      }
      return { payload: { action } };
    }
    case 'lockapp':
      if (!args[0]) {
        return { error: 'Usage: !lockapp <package>' };
      }
      return { payload: { packageName: args[0] } };
    case 'unlockapp':
      if (!args[0]) {
        return { error: 'Usage: !unlockapp <package>' };
      }
      return { payload: { packageName: args[0] } };
    default:
      return { error: 'Unknown remote command' };
  }
}

module.exports = {
  REMOTE_COMMANDS,
  ADMIN_COMMANDS,
  parseCommandInput,
  buildRemotePayload,
};

File: A-Dex/discord-bot/bot/command_parser.py
import re
from dataclasses import dataclass
from typing import Any

REMOTE_COMMANDS = {
    "apps",
    "open",
    "lock",
    "say",
    "sayurdu",
    "playaudio",
    "stopaudio",
    "pauseaudio",
    "resumeaudio",
    "audiostatus",
    "parentpin",
    "shield",
    "screenshot",
    "files",
    "filestat",
    "mkdir",
    "rename",
    "move",
    "delete",
    "uploadfile",
    "readtext",
    "download",
    "volume",
    "info",
    "permstatus",
    "location",
    "camerasnap",
    "contactlookup",
    "smsdraft",
    "fileshareintent",
    "quicklaunch",
    "torchpattern",
    "ringtoneprofile",
    "screentimeoutset",
    "mediacontrol",
    "randomquote",
    "fakecallui",
    "shakealert",
    "vibratepattern",
    "beep",
    "countdownoverlay",
    "flashtext",
    "coinflip",
    "diceroll",
    "randomnumber",
    "quicktimer",
    "soundfx",
    "prankscreen",
    "show",
    "message",
    "lockapp",
    "unlockapp",
    "lockedapps",
    "usage",
}

ADMIN_COMMANDS = {"pair", "bind", "unbind", "admins", "devices"}
TOKEN_REGEX = re.compile(r'"([^"]+)"|\'([^\']+)\'|(\S+)')


@dataclass(frozen=True)
class ParsedCommand:
    name: str
    args: list[str]
    is_remote: bool
    is_admin: bool



def tokenize(text: str) -> list[str]:
    """Tokenize command text with quoted argument support."""
    tokens: list[str] = []
    for match in TOKEN_REGEX.finditer(text):
        tokens.append(match.group(1) or match.group(2) or match.group(3))
    return tokens



def parse_command_input(content: str | None, prefix: str) -> ParsedCommand | None:
    if not content or not content.startswith(prefix):
        return None

    raw = content[len(prefix) :].strip()
    if not raw:
        return None

    tokens = tokenize(raw)
    if not tokens:
        return None

    name = tokens[0].lower()
    args = tokens[1:]
    return ParsedCommand(
        name=name,
        args=args,
        is_remote=name in REMOTE_COMMANDS,
        is_admin=name in ADMIN_COMMANDS,
    )



def build_remote_payload(name: str, args: list[str], attachment: dict[str, Any] | None, max_image_bytes: int) -> dict[str, Any]:
    if name in {
        "apps",
        "lock",
        "stopaudio",
        "pauseaudio",
        "resumeaudio",
        "audiostatus",
        "screenshot",
        "files",
        "camerasnap",
        "randomquote",
        "info",
        "permstatus",
        "location",
        "usage",
        "lockedapps",
        "coinflip",
    }:
        return {"payload": {}}

    if name == "open":
        if not args:
            return {"error": "Usage: !open <app package or display name>"}
        return {"payload": {"target": " ".join(args)}}

    if name == "say":
        if not args:
            return {"error": "Usage: !say <text>"}
        return {"payload": {"text": " ".join(args)}}

    if name == "sayurdu":
        if not args:
            return {"error": "Usage: !sayurdu <urdu text>"}
        return {"payload": {"text": " ".join(args)}}

    if name == "playaudio":
        if not args:
            return {"error": "Usage: !playaudio <url> [repeat]"}
        repeat = 1
        if len(args) > 1:
            try:
                repeat = int(float(args[1]))
            except ValueError:
                return {"error": "Usage: !playaudio <url> [repeat]"}
            if repeat < 1 or repeat > 100:
                return {"error": "Usage: !playaudio <url> [repeat]"}
        return {"payload": {"url": args[0], "repeat": repeat}}

    if name == "parentpin":
        if not args:
            return {"error": "Usage: !parentpin <4-12 digit pin>"}
        return {"payload": {"pin": args[0]}}

    if name == "shield":
        action = args[0].lower() if args else "status"
        if action not in {"enable", "disable", "status", "relock"}:
            return {"error": "Usage: !shield <enable|disable|status|relock>"}
        return {"payload": {"action": action}}

    if name == "download":
        if not args:
            return {"error": "Usage: !download <path>"}
        return {"payload": {"path": " ".join(args)}}

    if name == "filestat":
        if not args:
            return {"error": "Usage: !filestat <path>"}
        return {"payload": {"path": " ".join(args)}}

    if name == "mkdir":
        if not args:
            return {"error": "Usage: !mkdir <path>"}
        return {"payload": {"path": " ".join(args)}}

    if name == "rename":
        if len(args) < 2:
            return {"error": "Usage: !rename <path> <new_name>"}
        return {"payload": {"path": args[0], "new_name": args[1]}}

    if name == "move":
        if len(args) < 2:
            return {"error": "Usage: !move <source> <target_dir>"}
        return {"payload": {"source": args[0], "target_dir": args[1]}}

    if name == "delete":
        if not args:
            return {"error": "Usage: !delete <path> [recursive:true|false]"}
        recursive = False
        if len(args) > 1:
            recursive = args[1].lower() == "true"
        return {"payload": {"path": args[0], "recursive": recursive}}

    if name == "uploadfile":
        if len(args) < 2:
            return {"error": "Usage: !uploadfile <target_dir> <url> [file_name]"}
        payload = {"target_dir": args[0], "url": args[1]}
        if len(args) > 2:
            payload["file_name"] = args[2]
        return {"payload": payload}

    if name == "readtext":
        if not args:
            return {"error": "Usage: !readtext <path> [max_chars]"}
        max_chars = 2000
        if len(args) > 1:
            try:
                max_chars = int(float(args[1]))
            except ValueError:
                return {"error": "Usage: !readtext <path> [max_chars]"}
        return {"payload": {"path": args[0], "max_chars": max_chars}}

    if name == "volume":
        try:
            value = int(float(args[0]))
        except (IndexError, ValueError):
            return {"error": "Usage: !volume <0-100>"}

        if value < 0 or value > 100:
            return {"error": "Usage: !volume <0-100>"}
        return {"payload": {"value": value}}

    if name == "show":
        try:
            seconds = int(float(args[0]))
        except (IndexError, ValueError):
            return {"error": "Usage: !show <1-60> with an image attachment"}

        if seconds < 1 or seconds > 60:
            return {"error": "Usage: !show <1-60> with an image attachment"}

        if not attachment:
            return {"error": "Attach a single image to use !show"}

        content_type = (attachment.get("content_type") or "").lower()
        if not content_type.startswith("image/"):
            return {"error": "Attachment must be an image"}

        size = attachment.get("size")
        if isinstance(size, int) and size > max_image_bytes:
            return {"error": f"Attachment too large; max {max_image_bytes // (1024 * 1024)} MB"}

        return {
            "payload": {
                "seconds": seconds,
                "imageUrl": attachment.get("url"),
                "imageName": attachment.get("name") or "image",
                "imageContentType": content_type or "image/*",
            }
        }

    if name == "message":
        if not args:
            return {"error": "Usage: !message <text>"}
        return {"payload": {"text": " ".join(args)}}

    if name == "contactlookup":
        if not args:
            return {"error": "Usage: !contactlookup <query> [limit]"}
        payload: dict[str, Any] = {"query": args[0]}
        if len(args) > 1:
            try:
                payload["limit"] = int(float(args[1]))
            except ValueError:
                return {"error": "Usage: !contactlookup <query> [limit]"}
        return {"payload": payload}

    if name == "smsdraft":
        if len(args) < 2:
            return {"error": "Usage: !smsdraft <number> <message>"}
        return {"payload": {"number": args[0], "message": " ".join(args[1:])}}

    if name == "fileshareintent":
        if not args:
            return {"error": "Usage: !fileshareintent <path> [mime_type]"}
        payload: dict[str, Any] = {"path": args[0]}
        if len(args) > 1:
            payload["mimeType"] = args[1]
        return {"payload": payload}

    if name == "quicklaunch":
        if not args:
            return {"error": "Usage: !quicklaunch <package_or_url>"}
        target = args[0]
        if target.startswith("http://") or target.startswith("https://"):
            return {"payload": {"url": target}}
        return {"payload": {"packageName": target}}

    if name == "torchpattern":
        repeats = 3
        on_ms = 250
        off_ms = 250
        if len(args) > 0:
            try:
                repeats = int(float(args[0]))
            except ValueError:
                return {"error": "Usage: !torchpattern [repeats] [on_ms] [off_ms]"}
        if len(args) > 1:
            try:
                on_ms = int(float(args[1]))
            except ValueError:
                return {"error": "Usage: !torchpattern [repeats] [on_ms] [off_ms]"}
        if len(args) > 2:
            try:
                off_ms = int(float(args[2]))
            except ValueError:
                return {"error": "Usage: !torchpattern [repeats] [on_ms] [off_ms]"}
        return {"payload": {"repeats": repeats, "on_ms": on_ms, "off_ms": off_ms}}

    if name == "ringtoneprofile":
        if not args:
            return {"error": "Usage: !ringtoneprofile <normal|vibrate|silent>"}
        mode = args[0].lower()
        if mode not in {"normal", "vibrate", "silent"}:
            return {"error": "Usage: !ringtoneprofile <normal|vibrate|silent>"}
        return {"payload": {"mode": mode}}

    if name == "screentimeoutset":
        if not args:
            return {"error": "Usage: !screentimeoutset <seconds>"}
        try:
            seconds = int(float(args[0]))
        except ValueError:
            return {"error": "Usage: !screentimeoutset <seconds>"}
        return {"payload": {"seconds": seconds}}

    if name == "mediacontrol":
        if not args:
            return {"error": "Usage: !mediacontrol <play|pause|next|previous|stop|toggle>"}
        action = args[0].lower()
        if action not in {"play", "pause", "next", "previous", "stop", "toggle"}:
            return {"error": "Usage: !mediacontrol <play|pause|next|previous|stop|toggle>"}
        return {"payload": {"action": action}}

    if name == "fakecallui":
        caller_name = " ".join(args) if args else "Unknown Caller"
        return {"payload": {"callerName": caller_name}}

    if name == "shakealert":
        action = args[0].lower() if args else "status"
        if action not in {"start", "stop", "status"}:
            return {"error": "Usage: !shakealert <start|stop|status>"}
        return {"payload": {"action": action}}

    if name == "vibratepattern":
        if not args:
            return {"error": "Usage: !vibratepattern <comma_ms_pattern> [repeat:true|false]"}
        try:
            pattern = [int(v.strip()) for v in args[0].split(",") if v.strip()]
        except ValueError:
            return {"error": "Usage: !vibratepattern <comma_ms_pattern> [repeat:true|false]"}
        if not pattern:
            return {"error": "Usage: !vibratepattern <comma_ms_pattern> [repeat:true|false]"}
        repeat = len(args) > 1 and args[1].lower() == "true"
        return {"payload": {"patternMs": pattern, "repeat": repeat}}

    if name == "beep":
        tone = args[0].lower() if args else "beep"
        count = 1
        if len(args) > 1:
            try:
                count = int(float(args[1]))
            except ValueError:
                return {"error": "Usage: !beep [tone] [count]"}
        return {"payload": {"tone": tone, "count": count}}

    if name == "countdownoverlay":
        seconds = 10
        if args:
            try:
                seconds = int(float(args[0]))
            except ValueError:
                return {"error": "Usage: !countdownoverlay [seconds] [message]"}
        message = " ".join(args[1:]).strip() if len(args) > 1 else "Break over"
        return {"payload": {"seconds": seconds, "message": message}}

    if name == "flashtext":
        if not args:
            return {"error": "Usage: !flashtext <text> [seconds]"}
        seconds = 8
        text = " ".join(args)
        if args[-1].isdigit():
            seconds = int(args[-1])
            text = " ".join(args[:-1]).strip()
        if not text:
            return {"error": "Usage: !flashtext <text> [seconds]"}
        return {"payload": {"text": text, "seconds": seconds}}

    if name == "diceroll":
        sides = 6
        count = 1
        if len(args) > 0:
            try:
                sides = int(float(args[0]))
            except ValueError:
                return {"error": "Usage: !diceroll [sides] [count]"}
        if len(args) > 1:
            try:
                count = int(float(args[1]))
            except ValueError:
                return {"error": "Usage: !diceroll [sides] [count]"}
        return {"payload": {"sides": sides, "count": count}}

    if name == "randomnumber":
        minimum = 1
        maximum = 100
        if len(args) > 0:
            try:
                minimum = int(float(args[0]))
            except ValueError:
                return {"error": "Usage: !randomnumber [min] [max]"}
        if len(args) > 1:
            try:
                maximum = int(float(args[1]))
            except ValueError:
                return {"error": "Usage: !randomnumber [min] [max]"}
        return {"payload": {"min": minimum, "max": maximum}}

    if name == "quicktimer":
        seconds = 30
        label = "Timer"
        if len(args) > 0:
            try:
                seconds = int(float(args[0]))
            except ValueError:
                return {"error": "Usage: !quicktimer [seconds] [label]"}
        if len(args) > 1:
            label = " ".join(args[1:]).strip() or "Timer"
        return {"payload": {"seconds": seconds, "label": label}}

    if name == "soundfx":
        effect = args[0].lower() if args else "applause"
        duration_ms = 3000
        if len(args) > 1:
            try:
                duration_ms = int(float(args[1]))
            except ValueError:
                return {"error": "Usage: !soundfx [effect] [duration_ms]"}
        return {"payload": {"effect": effect, "durationMs": duration_ms}}

    if name == "prankscreen":
        mode = args[0].lower() if args else "glitch"
        seconds = 6
        if len(args) > 1:
            try:
                seconds = int(float(args[1]))
            except ValueError:
                return {"error": "Usage: !prankscreen [mode] [seconds]"}
        return {"payload": {"mode": mode, "seconds": seconds}}

    if name == "lockapp":
        if not args:
            return {"error": "Usage: !lockapp <package>"}
        return {"payload": {"packageName": args[0]}}

    if name == "unlockapp":
        if not args:
            return {"error": "Usage: !unlockapp <package>"}
        return {"payload": {"packageName": args[0]}}

    return {"error": "Unknown remote command"}
